docs/source/posts/2024/2024-08-21-final-report-robin.rst

```diff
@@ -0,0 +1,305 @@
+    Choosing the technology depended largely on HuggingFace hardware support. We experimented with inferencing LlamaCPP directly, inferencing Ollama, tested different quantizations and so on. Phi-3-mini-4k-instruct was chosen initially as the LLM. We rolled with it using Ollama for a few weeks. But as luck has it, I ended up discovering Groq is a cloud provider that provides free LLM endpoints. We used Groq from then on, and later also integrated Gemini since they also have a free tier.
+
+    You can hot-swap between a local model, a Groq model, a Gemini normal model or a Gemini Fine-Tuned model as you wish using the FURY Engine endpoint. it'll all integrate cleanly with the Pinecone database outputs and give a standard API response.
+
+
+- **Work on the backend architecture**
+    This is the present backend architecture.
+
+    .. image:: /_static/images/gsoc_llm_robin_week5.jpg
+        :alt: Present backend architecture
+
+
+    You're only hitting the FURY Engine endpoint, the remaining are all abstracted away. You can tell the engine you need to use Gemini and it'll do that for you. This is also expandable, if you have a new provider, you can add a new endpoint and connect it to FURY Engine.
+
+    The data to the REST endpoint will look like this
+
+    .. code-block:: json
+
+        {
+        "query": "Render a cube in fury",
+        "llm": "llama3-70b-8192",
+        "knn": "3",
+        "stream": false
+        }
+
+    Every output response will look like this
+
+    .. code-block:: json
+
+        {
+        "response": "Yes, this is how it would be done python import fury....",
+        "references": "1, 2, 3"
+        }
+
+    So if you do
+
+    `curl -X POST https://robinroy03-fury-engine.hf.space/api/groq/generate -H "Content-Type: application/json" -d '{"query": "How do I create a sphere in FURY?", "llm": "llama3-70b-8192", "knn": "3", "stream": false}'`
+
+    You'll get a response from ``llama3-70b-8192`` using ``Groq``. If you do ``https://robinroy03-fury-engine.hf.space/api/google/generate`` you can call any Google Gemini models like ``gemini-1.5-pro`` or ``gemini-1.5-flash``. Same for ``Ollama``.
+
+    A detailed blog on architecture is available `here. <https://fury.gl/latest/posts/2024/2024-07-01-week-5-robin.html>`_
+
+
+- **Work on improving model accuracy**
+    The initial version had major issues of hallucination and was unable to retrieve relevant context. We fix them by collecting more data, improving RAG, setting up a benchmark and so on.
+
+    The Initial version used a naive parser to parse code, later my mentors told me to use an AST parser. I chunked the entire repo using this and it performed relatively better. For model benchmarking, we had 2 tests, one QnA testing and one code testing. If the code compiles, the model gets one point.
+
+    All the benchmarking, data parsing, and database upsertion scripts are `here. <https://github.com/robinroy03/FURY-data-script>`_
```

Comment:
Using the same named references several times (“here” in our case) is not recommended with Sphinx, to avoid warnings you should Use anonymous hyperlinks, anonymous hyperlinks are denoted by the use of two underscores (__) instead of one. This way, each occurrence of the link text can have its own target.


``here <https://fury.gl/latest/posts/2024/2024-07-01-week-5-robin.html>``__ instead of ``here <https://fury.gl/latest/posts/2024/2024-07-01-week-5-robin.html>``_

Reply:
I didn't know this trick, it would've saved tons of time in my blogs... ty Wachiou

Reply:
@WassCodeur wow that's new for me, ty. Will update :+1: